compliance/collection/main.go

```diff
@@ -182,17 +217,106 @@ func manageNodeScanLoop(ctx context.Context, rescanInterval time.Duration, scann
+	inventory, err := opts.scanner.Scan(opts.nodeName)
 	if err != nil {
 		return nil, err
 	}
+	inv, err := proto.Marshal(inventory)
+	if err != nil {
+		return nil, err
+	}
+	if err := os.WriteFile(fmt.Sprintf("%s/last_scan", opts.inventoryCachePath), inv, 0600); err != nil {
+		return nil, err
+	}
+
+	return inventory, nil
+}
+
+func createAndObserveMessage(nodeName string, inventory *storage.NodeInventory) *sensor.MsgFromCompliance {
 	msg := &sensor.MsgFromCompliance{
 		Node: nodeName,
-		Msg:  &sensor.MsgFromCompliance_NodeInventory{NodeInventory: result},
+		Msg:  &sensor.MsgFromCompliance_NodeInventory{NodeInventory: inventory},
 	}
 	cmetrics.ObserveInventoryProtobufMessage(msg)
-	return msg, nil
+	return msg
+}
+
+// scanNodeWithBackoff runs scans with a linear backoff based on a file to not overstrain a Node if the container keeps restarting.
+// The backoff file will only be encountered if the previous container is killed during a call to scanNodeWithBackoff.
+// Note: This does not prevent strain in case of repeated pod recreation, as it is based on an EmptyDir.
+func scanNodeWithBackoff(opts *cachedScanOpts) (*sensor.MsgFromCompliance, error) {
+	backoffIntervalSeconds := int64(env.NodeInventoryInitialBackoff.DurationSetting() / time.Second)
+
+	backoffFile, err := os.ReadFile(fmt.Sprintf("%s/backoff", opts.inventoryCachePath))
+	defer func() {
+		if err := os.Remove(fmt.Sprintf("%s/backoff", opts.inventoryCachePath)); err != nil {
+			log.Warnf("Could not remove backoff state file: %v", err)
+		}
+	}()
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			log.Debug("No backoff found, continuing without pause")
+		} else {
+			return nil, err
+		}
+	} else {
+		// We have an existing backoff counter
+		backoffIntervalMillis, err := strconv.ParseInt(string(backoffFile), 10, 64)
+		if err != nil {
+			return nil, err
+		}
+		backoffIntervalSeconds = int64(time.Duration(backoffIntervalMillis) / time.Second)
+		log.Infof("Found existing backoff. Waiting before running next inventory")
+		backoffIntervalSeconds = opts.backoffWaitCallback(backoffIntervalSeconds)
+	}
+
+	err = os.WriteFile(fmt.Sprintf("%s/backoff", opts.inventoryCachePath), []byte(fmt.Sprintf("%d", backoffIntervalSeconds)), 0600)
+	if err != nil {
+		return nil, err
+	}
+
+	message, err := cachedScanNode(opts)
+	if err != nil {
+		return nil, err
+	}
+	return message, nil
+}
+
+// cachedScanNode checks for a cached inventory before running a new scan
+func cachedScanNode(opts *cachedScanOpts) (*sensor.MsgFromCompliance, error) {
+	var inventory *storage.NodeInventory
+
+	cachedInv, err := os.ReadFile(fmt.Sprintf("%s/last_scan", opts.inventoryCachePath))
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			log.Debug("No cache file found, running new inventory")
+		}
+		// Running into any other error than ErrNotExist continues to a fresh inventory
+		log.Infof("Unable to read inventory cache, running new inventory. Error: %v", err)
```

Comment:
I suggest it's worth a warning because something corrupted our cache.

Reply:
Done